# Natural Language Processing, Homework 1
 Dr. Sharon Yalov-Handzel

You can submit the homework as *.py or *.ipynb file.  
Answers to theoretical questions can be written in a separate WORD document.


Submitted by  
Michael Berger,  
Barak Simoni


  ## 1. Setup a python environment on your computer (Pycharm / google colab).



  ## 2. Install the following libraries: pytorch, NLTK, spaCy
  


In [82]:
#conda install -y pytorch NLTK spaCy chardet pandasql

Retrieving notices: ...working... done
Channels:
 - defaults
 - conda-forge
Platform: linux-64
Solving environment: done


==> WARNING: A newer version of conda exists. <==
    current version: 24.7.1
    latest version: 24.11.2

Please update conda by running

    $ conda update -n base -c conda-forge conda



## Package Plan ##

  environment location: /media/michael/Data/anaconda3/envs/NLP

  added / updated specs:
    - pandasql


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    greenlet-3.1.1             |  py311h6a678d5_0         243 KB
    sqlalchemy-2.0.34          |  py311h00e1ef3_0         6.6 MB
    ------------------------------------------------------------
                                           Total:         6.8 MB

The following NEW packages will be INSTALLED:

  greenlet           pkgs/main/linux-64::greenlet-3.1.1-py311h6a678d5_0 
  pandasql           pkgs/main/linux


  ## 3. Write a program that loads the file spam.csv (source: KAGGLE)


In [83]:
import pandas as pd
import pandasql as ps 
import numpy as np
from IPython.display import Markdown, display
import re
import chardet

In [51]:
file = "data/spam.csv"
spam = pd.read_csv(file ,encoding='Windows-1252')

In [52]:

with open(file, 'rb') as rawdata:
    result = chardet.detect(rawdata.read(100000))
result

{'encoding': 'Windows-1252', 'confidence': 0.7270322499829184, 'language': ''}

In [53]:
spam

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN
...,...,...,...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...,NaN,NaN,NaN
5568,ham,Will Ì_ b going to esplanade fr home?,NaN,NaN,NaN
5569,ham,"Pity, * was in mood for that. So...any other s...",NaN,NaN,NaN
5570,ham,The guy did some bitching but I acted like i'd...,NaN,NaN,NaN


In [54]:
spam[spam.iloc[:,2].notna()]

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
95,spam,Your free ringtone is waiting to be collected....,PO Box 5249,"MK17 92H. 450Ppw 16""",NaN
281,ham,\Wen u miss someone,the person is definitely special for u..... B...,why to miss them,"just Keep-in-touch\"" gdeve.."""
444,ham,\HEY HEY WERETHE MONKEESPEOPLE SAY WE MONKEYAR...,HOWU DOIN? FOUNDURSELF A JOBYET SAUSAGE?LOVE ...,NaN,NaN
671,spam,SMS. ac sun0819 posts HELLO:\You seem cool,"wanted to say hi. HI!!!\"" Stop? Send STOP to ...",NaN,NaN
710,ham,Height of Confidence: All the Aeronautics prof...,"this wont even start........ Datz confidence..""",NaN,NaN
899,spam,Your free ringtone is waiting to be collected....,PO Box 5249,"MK17 92H. 450Ppw 16""",NaN
1038,ham,"Edison has rightly said, \A fool can ask more ...",GN,GE,"GNT:-)"""
1127,ham,"Height of \Oh shit....!!\"" situation: A guy th...",".;-):-D""",NaN,NaN
1266,ham,\Hey sorry I didntgive ya a a bellearlier hunny,just been in bedbut mite go 2 thepub l8tr if u...,NaN,NaN
1384,ham,"Storming msg: Wen u lift d phne, u say \HELLO\...","bt not his girlfrnd... G o o d n i g h t . . .@""",NaN,NaN


Some messages got spread into several columns of the dataframe  
This happened due to inherent structure of messages, which contain all the possible characters,  
The presence of " and , chars in the messages makes separating tricky.  

We can solve it easily by concatenating the columns back into one column

In [56]:
def empty_join(lst, sep = ", "):
  return sep.join([x for x in lst if x])

In [57]:
spam.columns = "v1,v2,v3,v4,v5".split(",")
spam["msg"] = spam[['v2', 'v3', 'v4', 'v5']].fillna('').agg(empty_join, axis=1)
spam[["v1","msg"]]

,v1,msg
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will Ì_ b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...



  ## 4. Compute simple statistics on the data:   


In [58]:

rows = spam.msg.apply(func=lambda n: re.split("[\d\W]+", n.lower()))
words = sum(rows.values.tolist(),[])

wordsPerMess = rows.apply(func = len)

unique, counts = np.unique(words, return_counts=True)
wc = pd.DataFrame({"token":unique, "count": counts})

tokens = wc.sort_values("count",ascending=False)

In [60]:
show = rows[:20].apply(lambda r:  " ".join(r))
print("\n".join(show))

go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat 
ok lar joking wif u oni 
free entry in a wkly comp to win fa cup final tkts st may text fa to to receive entry question std txt rate t c s apply over s
u dun say so early hor u c already then say 
nah i don t think he goes to usf he lives around here though
freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send å to rcv
even my brother is not like to speak with me they treat me like aids patent 
as per your request melle melle oru minnaminunginte nurungu vettam has been set as your callertune for all callers press to copy your friends callertune
winner as a valued network customer you have been selected to receivea å prize reward to claim call claim code kl valid hours only 
had your mobile months or more u r entitled to update to the latest colour mobiles with camera for free call the mobile update co free on 

In [61]:
ans = f"""
number of SMS messages: {spam.shape[0]},  
number of spams: {sum(spam.v1=="spam")},   
total word count: {len(words)},   
averaged number of words per message: {wordsPerMess.mean():.2f},   
5 most frequent words in the file: {tokens.token.head(5).values},   
number of rare words (appear only once in the file): {sum(tokens["count"] == 1)}.  
"""

display(Markdown(ans))


number of SMS messages: 5572,  
number of spams: 747,   
total word count: 91723,   
averaged number of words per message: 16.46,   
5 most frequent words in the file: ['' 'i' 'to' 'you' 'a'],   
number of rare words (appear only once in the file): 3891.  



  ## 5. Perform tokenization twice: 


  ### a) by using NLTK library 
  
Barak

In [ ]:
from nltk.tokenize import word_tokenize
spam['nltk_tokens'] = spam['msg'].apply(lambda text: word_tokenize(text))

  ### b) and by Spacy library. 

Michael

In [62]:
#! python -m spacy download en_core_web_sm

In [ ]:
import spacy
nlp = spacy.load("en_core_web_sm")

spam['spacy'] = spam['msg'].apply(nlp)

Take result of SpaCy nlp pileline as plain list of str - for comparison

In [103]:
spam["spacy_tokens"] = spam.spacy.apply(lambda row: [r.text for r in row])
spam.spacy_tokens

0       [Go, until, jurong, point, ,, crazy, .., Avail...
1                [Ok, lar, ..., Joking, wif, u, oni, ...]
2       [Free, entry, in, 2, a, wkly, comp, to, win, F...
3       [U, dun, say, so, early, hor, ..., U, c, alrea...
4       [Nah, I, do, n't, think, he, goes, to, usf, ,,...
                              ...                        
5567    [This, is, the, 2nd, time, we, have, tried, 2,...
5568    [Will, Ì, _, b, going, to, esplanade, fr, home...
5569    [Pity, ,, *, was, in, mood, for, that, ., So, ...
5570    [The, guy, did, some, bitching, but, I, acted,...
5571                  [Rofl, ., Its, true, to, its, name]
Name: spacy_tokens, Length: 5572, dtype: object

### c) Write what is the difference between the results.

We will start by finding the different tokens between the 2 tokenizators

In [ ]:
spam['token_diff'] = spam.apply(lambda row: set(row['spacy_tokens']).symmetric_difference(set(row['nltk_tokens'])), axis=1)

In [208]:
df = spam[spam.token_diff!=set()]
df.shape

(1851, 13)

Out of 5571 messages, only 1851 (33%) got tokenized differently.  
Let's check out what these differences are.

We will output the tokenisations for these messages into 2 separate text files and visualize their difference using a standard OS visual diff tool

In [ ]:
df.spacy_tokens.apply(" ".join).to_csv("spacy_tokens.csv")

In [ ]:
df.nltk_tokens.apply(" ".join).to_csv("nltk_tokens.csv")

![](tokenisation_diff.png)

Upon examining the diff we can conclude that most of the diffences are in the treatment of the punctuation marks.  
Most notably in non-standard use of them due to slang, informal writing and mistakes, characteristic to SMS message communication    


We can investigate this further.  
Let's examine individual tokens that get tokenized differently

In [210]:

spam['spacy_only'] = spam.apply(lambda row: list(set(row['spacy_tokens']).difference(set(row['nltk_tokens' ]))), axis=1)
spam['nltk_only']  = spam.apply(lambda row: list(set(row['nltk_tokens' ]).difference(set(row['spacy_tokens']))), axis=1)


In [175]:
so = spam.explode('spacy_only')[['spacy_only']]

q = """
select spacy_only, count(spacy_only) as cnt
from so 
group by spacy_only
order by cnt desc
limit 50
"""

so = ps.sqldf(q).head(50)

,spacy_only
0,NaN
1,NaN
2,rate)T&C
2,question(std
3,NaN
...,...
5568,Ì
5568,_
5569,NaN
5570,NaN


In [ ]:
no = spam.explode('nltk_only')[['nltk_only']]

q = """
select nltk_only, count(nltk_only) as cnt
from no 
group by nltk_only
order by cnt desc
limit 50
"""

no = ps.sqldf(q).head(50)

,nltk_only
0,NaN
1,NaN
2,rate
2,std
2,&
...,...
5567,BT-national-rate
5568,Ì_
5569,NaN
5570,NaN


And so we acquire 2 lists of tokens most frequently being tokenized differently between the two libraries:

In [187]:
together = pd.concat([no,so], axis=1)
together

,nltk_only,cnt,spacy_only,cnt
0,:,345,,302
1,),244,nt,291
2,lt,239,lt;#&gt,214
3,gt,234,/,170
4,#,212,:),157
5,'',143,"""",145
6,-,114,-,135
7,Ì_,97,_,111
8,dont,96,Ì,95
9,&,78,do,90


Most of them really appear to be caused by punctuation 


  ## 6. Perform lemmatization 
  by NLTK and  
  barak


  by Spacy libraries. 

Michael

In [80]:
for w in spam.spacy[4]:
  print(w, " | ", w.text, " | " ,w.lemma_)


Nah  |  Nah  |  nah
I  |  I  |  I
do  |  do  |  do
n't  |  n't  |  not
think  |  think  |  think
he  |  he  |  he
goes  |  goes  |  go
to  |  to  |  to
usf  |  usf  |  usf
,  |  ,  |  ,
he  |  he  |  he
lives  |  lives  |  live
around  |  around  |  around
here  |  here  |  here
though  |  though  |  though


  Write what is the difference between the results.




  ## 7. Perform stemming 
  by NLTK twice on both results of (6). Write what is the difference between the results.

Barak


  ## 8. Identify a spam message where its removal from the spam.csv dataset would:




  ### a)Reduce the total number of stemmed tokens
  
BArak


  ### b)Maintain the exact same number of lemmatized tokens
Michael

If no such message exists, provide a detailed explanation of why this scenario is
impossible or improbable.




  ## 9. Identify a spam message where its removal from the spam.csv dataset would:



  ### a)Reduce the total number of lemmatized tokens
  
Barak


  ### b)Maintain the exact same number of stemmed tokens

  Michael
If no such message exists, provide a detailed explanation of why this scenario is
impossible or improbable.




  ## 10. Compare the results of  
  NLTK before and after tokenization, lemmatization and stemming

  BArak



  ## 11. Compare the results of 
  Spacy before and after tokenization and lemmatization.

MIchael


  ## 12. Install the library BeautifulSoup



  ## 13. Choose any url page. 
  You can choose a page of yourself in a social media (e.g., facebook, Instagram, Linkdin, Twitter), or anything else



  ## 14. Perform data scrapping of this url. 
  You can learn how to do it from the following links:  
https://www.edureka.co/blog/web-scraping-with-python/  
https://zindi.africa/learn/a-beginners-guide-to-scraping-data-from-social-media  
or, ask CHATGPT to program it for you.  

Michael


  ## 15. Perform steps 5-11 on the text achieved in (14).



  ## 16. Download one of your whatsapp chats into *.txt file 
  (at least 50 messages in Hebrew).

  Barack



  ## 17. Repeat steps 5-11 on the text file achieved in (16)
